Prevent possible blob data corruption after append cancellation

spodskubka opened the pull request:
A test demonstrating the problem is contained in the first commit. If it is run without the fixes in the following commits, the bug can be reproduced and debugged step-by-step.

The following commits replace the `UploadOffset` metadata element, which could get out-of-sync with the actual blob length, with direct use of the blob Content-Length property.

As for the `UploadOffset` metadata element, I had four possible choices regarding in-progress uploads that were created before this pull request and other developers' expectations of its existence (ordered from my least to most favorite choice):
1. keep setting it during create/append actions, but never actually use it in code => it could get out of sync, like before, but that would be temporary and it would do no harm anymore
2. completely remove it from code => existing in-progress uploads would be stuck with an unchanging value in the metadata element, which could cause confusion
3. keep its name/key in code (for now) and set it to a hint with the next append action => would tell developers where to find the value
4. keep its name/key in code (for now) and remove it automatically with the next append action => could developers be asking themselves, where it went?

In the end, I went with choice 3. If you prefer 2 (cleanest code) or 4 (cleanest blobs), I could switch to one of them. Choice 1 is out of the question, I think.

Fixes #12

spodskubka committed:
Add test demonstrating corruption when cancelling AppendDataAsync

spodskubka committed:
Refactor blob metadata reading in preparation for upload offset change

spodskubka committed:
Replace upload offset metadata with blob content length

spodskubka committed:
Remove duplicate code and optimize code

giometrix commented:
Thanks for your thoughtful response.

I *think* I prefer option 2; I understand the desire to not confuse devs, but ultimately I think the number of in-progress uploads is small, and the tax of adding the obsolete message on every file uploaded is high. I think adding a note in the readme of what happened to the field may be good enough. 

How strongly do you feel about keeping the obsolete field? I told you my preference, but I'm not married to option 2.

Side note: I wonder if uploadlength is relatively new/behavior changed. Weird that I didn't  use it. Either way, nice work!

spodskubka commented:
Option 2 is totally fine for me. I prefer clean code whenever possible, just thought it could confuse people and wasn't sure about your policy in this regard.

But I guess you are right that most people won't even notice this change. And if they do and are confused, an explanation in the README will clear that up.

As far as I can tell from the docs, the Content-Length property has always been there. After all, it is a standard HTTP response header and an important characteristic of every blob.

spodskubka committed:
Remove obsolete upload offset metadata key

spodskubka committed:
Add version notes to README explaining the upload offset changes

spodskubka commented:
I implemented option 2 and added some content to the README file. Please feel free to modify the text so that it suits you.

As you can see, I took the liberty to introduce a major version bump to 3.0.0 because of breaking changes (caused by every option except 1), which are also explained in the README.

giometrix commented on README.md:
```diff
@@ -30,4 +30,26 @@ so this extension was disabled.  If you care about this feature and are ok with
+
+#### Can I prevent blob data corruption with versions < 3.x?
+
+Yes, by choosing a TUS chunk size ≤ 4 MiB (4 194 304 bytes). Please note that such small chunk sizes are not recommended because they reduce upload performance.
```
I want to make sure I understand, as I haven't been in the code too much recently. I see this [const](https://github.com/giometrix/Xtensible.TusDotNet.Azure/blob/8d45119247e33f5560463cbb58806e6b2c1ef917/src/Xtensible.TusDotNet.Azure/AzureBlobTusStore.cs#L28) and I don't see a way of overriding (just did a cursory search though).

1. Clearly you had issues, so is it true that the issue only occurs with larger chunks (if they are indeed not configurable)?
2. "Please note that such small chunk sizes are not recommended" , again I may be missing the place where we make chunk sizes configurable.... but I didn't look to closely, just taking a cursory look. If I am correct should we open this up (as not part of this PR)?

To be clear, I do think making the change to use ContentLength is the right thing to do; I just want to make sure that I understand what is being said here and possibility to open up features.

spodskubka replied:
When a TUS client is uploading a file it does so in chunks. It could be a single chunk containing the whole file or the file could be split into multiple chunks, which are uploaded separately. Especially when resuming uploads, the chunk size is not the whole file but the remaining file size at most. Each chunk is sent as a HTTP PATCH request from the TUS client to the TUS server.

The chosen TUS chunk size actually depends on the TUS client implementation and configuration, as well as the TUS server limits.

The `ITusStore.AppendDataAsync(...)` method receives a stream of data, which is sent by the TUS client. The length of this stream is the TUS chunk size, that was chosen by the TUS client.

Now, the `AzureBlobTusStore` performs its own splitting of the data, because Azure Append Blobs can only be appended to in blocks. The maximum size of each block is defined by the `AppendBlobBlockSize` as 4 MiB, which were chosen, according to the README, to support checksum validation. (Honestly, I have not yet tried to understand the checksum code yet, so I can't say anything in that regard.)

To answer the actual questions:

1. Actually, the risk of the issue happening is not zero when using a TUS chunk size ≤ 4 MiB. Something (CancellationToken triggered, App killed, Azure connection interrupted) could prevent the UploadOffset Azure Metadata property update from happening after the actual data was appended to the Azure Blob. So the README text should actually read "You cannot fully prevent it, only minimize the risk by ...".
2. This library cannot influence the TUS chunk size in any way. It has to live with the data that arrives at the `AppendDataAsync(...)` method. As explained above, the TUS chunk size is chosen by the TUS client. The README text could be altered to make that clear. I suggest: "... by limiting your TUS client(s) to a chunk size ≤ 4 MiB, in that case it is also suggested to configure tusdotnet to disallow larger chunks.".

spodskubka replied:
@giometrix Oh, you already merged this without my suggested changes to the README. Should I submit another pull request?

giometrix replied:
You can if you want, and I’ll merge it.

I didn’t because I don’t think we should make too much of a big deal about upgrading to 3.0; I want to encourage everyone to upgrade, as it’s pretty painless and solves a potential corruption issue many may not have been aware of.

But if you feel like it’s an important note I don’t mind adding it.
